node/node-sdk.md

```diff
@@ -0,0 +1,187 @@
+```typescript
+import { ActivityContext } from 'temporal-sdk';
+import axios from 'axios';
+
+export async function httpGet(ctx: ActivityContext, url: string) {
+  const response = await axios.get(url, {
+    onDownloadProgress: (progressEvent) => {
+      const total = parseFloat(progressEvent.currentTarget.responseHeaders['Content-Length']);
+      const current = progressEvent.currentTarget.response.length;
+      const progress = Math.floor(current / total * 100);
+      ctx.heartbeat(progress);
+    },
+  });
+  return response.data;
+}
+```
+
+### Registration
+Activities should be registered in their corresponding workers.
+
+```typescript
+worker.registerActivity('httpGet', httpGet, { passContext: true });
+```
+
+### Payloads
+Initially we'll support any JSON serializable object as payload, binary data support can be added if required.
+
+TBD: consider supporting [typed arrays][typed-arrays].
+
+## Workflows
+Worflow code looks like vanilla TypeScript.
+Workflows run in the same NodeJS process as the Temporal SDK and the activities.
+Workflow code must export a `main` function which will be used as the handler method for workflow invocations.
+
+TODO: signal and query handlers.
+
+Each workflow instance runs in a separate [V8 isolate][v8-isolate], which means it has a completely isolated execution context including global variables.
+The Temporal SDK injects and removes references from the global object to ensure that code running in the isolate is fully deterministic.
+
+Activities are invoked by calling `activities.$activityName` as a method or with options (see below).
+
+### Example
+```typescript
+export async function main(x: number, y: number) {
+  const msToSleep = await activities.httpGet(`https://calculator.com/add?x=${x}&y={y}`);
+  await new Promise((resolve) => setTimeout(resolve, msToSleep * Math.random()));
+}
+```
+
+### Activity Options
+Activity options are passed in using the `withOptions` method.
+
+```typescript
+export async function main() {
+  const body = await activities.httpGet.withOptions({ retries: 3 }).invoke('http://example.com');
+  console.log(body);
+}
+```
+
+### Activity Type Declarations
+To define type declarations for your workflow's activities, add each activity to your `global.d.ts`.
+```typescript
+/// <reference types="@temporal-sdk/workflow-global" />
+
+import * as declarations from '../../activities';
+
+declare global {
+  namespace activities {
+    // Use the signature of an activity function
+    export var httpGet: Activity<typeof declarations.httpGet>;
+    // In case you can't import you function definition (implmented in another language or simply unavailable in this context),
+    // define the signature manually
+    export var add: Activity<Fn<[number, number], number>>;
+  }
+}
+```
+
+### Registration
+```typescript
+async function main() {
+  const example = 'dist/workflow.js'; // Workflow file generated using the typescript compiler
+  await worker.registerWorkflow(example);
```

Comment:
I'm confused. In the example above the workflow code was defined directly in the main method. But here we use the main method to initialize the worker.

Reply:
After taking another look I understand the confusion.
This `main` is in the entry point of the worker process not the workflow, I'll fix the example.